Try adding global scroll animations to Dawn

kjellr opened the pull request:
### PR Summary: 

This PR adds basic reveal animations to Dawn. 

Here's how it looks: 

https://user-images.githubusercontent.com/1202812/218783665-797dd11d-5044-4ac7-bebf-6b5b1c7a9d79.mp4

### What approach did you take?

The reveal animations are turned on globally via a single checkbox in Theme Settings. The PR uses a light Javascript `IteractionObserver` to add a `scrolled-into-view` class to various elements. This triggers a light CSS animation. 

Some additional notes: 

- The JS was lifted from @stufreen's [stu/scroll-trigger](https://github.com/Shopify/dawn/compare/stu/scroll-trigger) branch.
- It respects [prefers-reduced-motion](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion), and turns itself off if the `no-js` class is present.
- Some elements (cards, for example) are set to load in sequentially. ~~The `nth-child()` logic could use some close 👀 — every now and then the load order isn't 100% left-to-right as expected.~~ (Fixed in https://github.com/Shopify/dawn/pull/2141/commits/514e3f0909f42c6d4a06f8873d6d5b9a2daebf51)   
- I couldn't figure out where exactly to place the "Motion" item sequentially within Theme Settings — the order there feels a bit random

Followup tasks: 

- [ ] In general, some bug hunting. 
- [x] Test with no Javascript
- [ ] In the editor, new sections appear invisible when they're added in. 
- [ ] Test with an eye for potential special cases. (For example, do we want to modify the animation for a specific section?)

### Visual impact on existing themes

None at this point — this is currently built as an opt-in. We should re-visit this though before launch. 

### Testing steps/scenarios

1. Visit [the demo store](https://os2-demo.myshopify.com/admin/themes/137469296662/editor).
2. Visit Theme Settings > Motion
3. Check the "Reveal sections on scroll" checkbox.
4. Scroll around and find bugs! 

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

clauderic commented on assets/animations.js:
```diff
@@ -0,0 +1,15 @@
+(function() {
+    window.addEventListener('scroll', animateIn);
+    animateIn();
+    function animateIn() {
+        var elements = document.querySelectorAll('.section, .banner__box, .product__info-wrapper');
+        for (var i = 0; i < elements.length; i++) {
+            elements[i].classList.add('animate');
+            var windowHeight = window.innerHeight;
+            var scrollTop = elements[i].getBoundingClientRect().top;
+            if (scrollTop < windowHeight - 100) {
+                elements[i].classList.add('animate--active');
+            }
+        }
+    };
+})();
```
Scroll event listeners aren't very performant, especially on mobile browsers where they are throttled. Thankfully, there is an alternative! The [Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) was introduced for this very use case, and is widely supported in modern browsers.

You should be able to achieve something similar like so:

```ts
function onIntersection(entries) {
  for (const entry of entries) {
    // refer to https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserverEntry
    if (entry.isIntersecting) {
      entry.target.classList.add(...);
    }
  }
}

const intersectionObserver = new IntersectionObserver(onIntersection, {
  thresholds: [1], // refer to https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver/thresholds
});
const selector = '.section';
const elementsToObserve = Array.from(document.querySelectorAll(selector));

// Observe intersections between the viewport and the elements
elementsToObserve.forEach((element) => intersectionObserver.observe(element));
```

There's probably a few other things that may be worth handling, like listening to the [`shopify:section:load` event in the theme editor](https://shopify.dev/themes/architecture/sections/integrate-sections-with-the-theme-editor#javascript-events) to observe newly added sections that weren't there when the code executed for the first time:

```ts
document.addEventListener('shopify:section:load', (event) => {
  intersectionObserver.observe(target);
});
```

clauderic reviewed (commented):
Love the overall effect! I just have a comment on the approach from a technical perspective.

On the overall UX: on one hand, I love the simplicity of having a single on/off setting. I don't want to open a can of worms, but I can't help but wonder if we should offer any other options. I've seen a lot of sites that re-play the animation when the section has scrolled out of the viewport and is scrolled back into view. 

Things we could consider opening up:

- Type of animation (fade in, fade up, fade down, etc.)
- Whether animations should play only once or every time they appear into the viewport

I also think this would be a good use-case for overrides in the future.

ludoboludo commented:
Some cool things that we could look into using when they become available and widely supported: 
- [view transition API](https://developer.chrome.com/docs/web-platform/view-transitions/)
- [CSS scroll timeline](https://css-tricks.com/practical-use-cases-for-scroll-linked-animations-in-css-with-scroll-timelines/)

kjellr commented:
> On the overall UX: on one hand, I love the simplicity of having a single on/off setting. I don't want to open a can of worms, but I can't help but wonder if we should offer any other options. I've seen a lot of sites that re-play the animation when the section has scrolled out of the viewport and is scrolled back into view.

Yeah, I agree there's lots of room for additional variations here. But I think it makes sense to focus in on what we think the best default experience is, build that in, and then scale up with more variants in the future. That'll get us to launch quicker, and fulfill the needs of most merchants who normally wouldn't bother digging into settings for this sort of thing. 

> Some cool things that we could look into using when they become available and widely supported...

Very cool, @ludoboludo! Definitely good to keep an eye on those features. 🌟

stufreen commented on assets/base.css:
```diff
@@ -2957,3 +2957,41 @@ details-disclosure > details {
+/* Animations */
+
+.animate {
+  transition: opacity 0.55s ease-in-out;
```
You can experiment with the `cubic-bezier` easing function instead of `ease-in-out` if you want more control. This web app is a neat tool I just found: https://cubic-bezier.com/#.17,.51,.31,1

kjellr commented:
A few new updates yesterday: 

- Fixed `no-js` behavior.
- Fixed `nth-child` rules so things more reliably load in left-to-right.
- Added light animation to a bunch of Sections I’d missed the first time around.
- Set slideshow content to borrow the slide-up animation when slides are revealed. We can fine-tune these, but they're working now. 

https://user-images.githubusercontent.com/1202812/219448819-ac374973-92d1-4bf9-baeb-df2c4a0ce2a9.mp4

melissaperreault reviewed (commented):
Thank you Kjell! I took a first pass and here are some observations to help scoping and forming additional opinions for revealing sections on scroll combined to page load experience:

**Bugs**
- [ ] Product template: The product information section does not display. Same on the Quick add modal!
- [ ] Password template: When you add sections, they don't display on the page.
- [ ] Mobile: Could we delay revealing the section below the first one? ([Video reference](https://screenshot.click/20-41-75of6-xxasg.mp4))

**Nice to have**
- [ ] Product template: We could fade in cascade the media grid, than the product information column? There is something to explore here for more delight as you land on that page.
- [ ] Footer: I would recommend to keep the reveal fade in but not in cascade. I wonder if it is interesting for that section to reveal in cascade, because the section is at the end of the whole experience. When I experienced it for the first time, I felt overwhelmed because my discovery was almost done and I was proposed with some engaging motion in a space that I didn't expect delight. We could argue that this section is more informationally driven and requires less vibrance?
- [ ] Cart
   - [ ] Empty cart: We could fade in the content of the page to keep the reveal going for an homogeneous expression. ([Video reference](https://screenshot.click/20-17-vr230-tmqr5.mp4))
   - [ ] [Add to cart](https://screenshot.click/20-22-v4f2u-amc98.mp4): Popup notification could reveal more smoothly
   - [ ] Cart page with items: Related to the point above, when removing an item from your cart, the cart item line could disappear more smoothly.
- [ ] [Blog posts template](https://screenshot.click/20-31-q6x13-tdkkc.mp4): We offer 2 layout, Collage and Grid, we could animate in cascade here as well. 
- [ ] Pages, Blog post template (Article): To continue a seamless experience, the reveal impression would be interesting on those templates. Otherwise the content feels static surrounded by softer reveal of other sections. 
- [ ] [Collection template](https://screenshot.click/20-28-qwu1k-yiyox.mp4): Page load fade in for the title/description and Collection banner image
- [ ] Customer accounts: Page load and revealing elements in cascade perhaps?
- [ ] Password template: 
   - [ ] 1. We can reuse the same intention from the `Email signup` section to the `Email signup banner` [section](https://screenshot.click/20-05-iicci-zye6d.png). 
   - [ ] 2. We could fade in the content of the modal to remain in the same spirit. ([Video reference](https://screenshot.click/20-07-d44op-aq35a.mp4) of the component)

stufreen commented on assets/animations.js:
```diff
@@ -0,0 +1,31 @@
+    scrollTriggerElements.forEach((element) => observer.observe(element));
+}
+
+window.addEventListener("load", initializeScrollTrigger);
```
I added this - my thinking is that we should wait for images to load before adding the listeners, but I didn't test too thoroughly.

stufreen commented on assets/animations.js:
```diff
@@ -0,0 +1,31 @@
+    }
+}
+
+function initializeScrollTrigger() {
```
Good call - we have a separate ticket to add this observer and I included it as a requirement: https://github.com/Shopify/dawn/issues/2316

There's a script in Dawn called "theme-editor.js" that gets loaded in the editor, so we'll probably put it in there.

kjellr commented:
>  Product template: The product information section does not display. Same on the Quick add modal!

I'm unable to reproduce either of these issues. Can you share steps? 

<details>
  <summary>📹 Here's what I'm seeing</summary>

https://user-images.githubusercontent.com/1202812/220392151-d35d00de-652c-483c-8060-072a612546ba.mp4

</details>



> Password template: When you add sections, they don't display on the page.

Yeah, this is currently the case with any section. 😄 This will get solved in https://github.com/Shopify/dawn/issues/2316. 


> Mobile: Could we delay revealing the section below the first one? (Video reference)

Maaaaaybe. I think the logic might get a little complicated if we do sequential loading on a section level. But I'll look into it. 🤔

clauderic commented on assets/animations.js:
```diff
@@ -0,0 +1,31 @@
+const IN_VIEW_CLASSNAME = "scrolled-into-view";
+
+const OPTIONS = {
+    threshold: 0.5,
```
This comment applies throughout the whole file, not just here 😅

kjellr replied:
😅 Ha, whoops. Should be better now.

melissaperreault commented:
> I'm unable to reproduce either of these issues. Can you share steps?

Oh, that's a wild one! Just checked back and this is how I got to experience this:

- I am testing with [this Editor link](https://os2-demo.myshopify.com/admin/themes/137469296662/editor)
- [Product template](https://screenshot.click/21-33-688sp-3lbgn.mp4): Resolution seems to be around 1140px~. I am able to reveal the product information section on mobile, than it's visible on desktop.
- [Quick add modal](https://screenshot.click/21-37-pzu9f-790c3.mp4): Same as above, the modal does not display the product information

kjellr commented:
Oh wow, I see it now. Ok cool, that's 100% a bug. 😄
